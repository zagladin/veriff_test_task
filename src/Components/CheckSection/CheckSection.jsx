import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import className from 'classnames/bind';
import styles from './CheckSection.module.scss';
import CheckButton from './Check/CheckButton';

const CheckSection = (
  {
    description,
    onClick,
    disabled,
  },
) => {
  const cx = className.bind(styles);

  const sectionStyles = cx({
    checkButtonWrapper: true,
    disabled,
  });

  const [selectedState, setSelectedState] = useState(null);

  const handleClick = useCallback((value) => {
    setSelectedState(value === 'Yes');
    onClick(value);
  }, [disabled, onClick]);

  const handleDigits = useCallback((event) => {
    if (event.key === '1') {
      handleClick('Yes');
    }
    if (event.key === '2') {
      handleClick('No');
    }
  }, [handleClick]);

  const handleArrows = useCallback((event) => {
    const nextElement = event.currentTarget?.nextSibling;
    const previousElement = event.currentTarget?.previousSibling;

    if (event.key === 'ArrowDown') {
      return nextElement && nextElement.focus();
    }
    if (event.key === 'ArrowUp') {
      return previousElement && previousElement.focus();
    }
    return null;
  }, []);

  const checkKeyDown = useCallback((event) => {
    if (disabled) {
      handleArrows(event);
      return null;
    }

    handleArrows(event);
    handleDigits(event);
  }, [disabled, handleArrows, handleDigits]);

  useEffect(() => setSelectedState(null), [disabled]);

  // the workaround to fill the unselected checks with nulls, to send all data,
  // but be able to distinguish between selected and unselected.
  // Probably we can do it in the upper scope
  useEffect(() => {
    if (disabled) {
      onClick(null);
    }
  }, []);

  return (
    <div
      className={sectionStyles}
      onKeyDown={checkKeyDown}
      tabIndex={0}
    >
      <div className={styles.description}>
        {description}
      </div>
      <div className={styles.buttons}>
        <CheckButton
          onClick={() => handleClick('Yes')}
          disabled={disabled}
          text="Yes"
          active={selectedState === true}
        />
        <CheckButton
          onClick={() => handleClick('No')}
          disabled={disabled}
          text="No"
          active={selectedState === false}
        />
      </div>
    </div>
  );
};

CheckSection.propTypes = {
  description: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  disabled: PropTypes.bool.isRequired,
};

export default React.memo(CheckSection);
