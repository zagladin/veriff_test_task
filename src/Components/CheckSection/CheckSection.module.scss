.checkButtonWrapper {
  width: 100%;
  height: auto;
  min-height: 100px;
  display: flex;
  flex-direction: column;

  &:hover, &:focus {
    background-color: var(--highlight-color);
    cursor: pointer;
  }

  .description {
    font-size: 15px;
    line-height: 120%;
    color: #2b2d2d;
    margin-bottom: 12px;
    text-align: left;
  }

  .buttons {
    display: flex;
    width: 100%;
    max-width: 300px;
    height: auto;

    button:first-child {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }

    button:last-child {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
  }
}

.disabled {
  filter: grayscale(1);

  &:hover, &:focus {
    background-color: #FFF;
    cursor: auto;
  }
}


