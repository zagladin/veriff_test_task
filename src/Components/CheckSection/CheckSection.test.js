import React from 'react';
import renderer from 'react-test-renderer';
import CheckSection from './CheckSection';

const mockData = {
  onClick: () => console.log('Click section'),
  description: 'Test check section',
  disabled: false,
};

it('should match the snapshot', () => {
  const checkSection = renderer
      .create(
          <CheckSection
              description={mockData.description}
              disabled={mockData.disabled}
              onClick={mockData.onClick} />
      )
      .toJSON();
  expect(checkSection).toMatchSnapshot();
});

