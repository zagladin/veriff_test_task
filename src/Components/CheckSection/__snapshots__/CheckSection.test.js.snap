// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="checkButtonWrapper"
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="description"
  >
    Test check section
  </div>
  <div
    className="buttons"
  >
    <button
      className="checkButton inactive"
      disabled={false}
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="checkButton inactive"
      disabled={false}
      onClick={[Function]}
    >
      No
    </button>
  </div>
</div>
`;
